torch/csrc/jit/tensorexpr/ir_printer.cpp

```diff
@@ -119,23 +119,38 @@ void IRPrinter::visit(const CompareSelect* v) {
   os() << ")";
 }
 
+void formatFPSuffix(std::ostream& os, double v) {
+  // No suffix for doubles.
+}
+
+template<typename T>
+void formatFPSuffix(std::ostream& os, T v) {
+  os << (v == std::ceil(v) ? ".f" : "f");
+}
+
+template<typename T,
+         std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
+static void formatImm(std::ostream& os, T v) {
+  if (std::isnan(v)) {
+    os << "NAN";
+  } else if (std::isinf(v)) {
+    os << (v > 0 ? "POS_INFINITY" : "NEG_INFINITY");
+  } else {
+    os << std::setprecision(16) << v;
+    formatFPSuffix(os, v);
+  }
+}
+
+template<typename T,
+         std::enable_if_t<!std::is_floating_point<T>::value>* = nullptr>
+static void formatImm(std::ostream& os, T v) {
+  os << v;
+}
 
 #define IMM_PRINT_VISIT(Type, Name)           \
```

Comment:
Can this macro just go away?

Reply:
Hm, maybe.  We could manually expand all the `void visit(const FooImm*)` functions, or template the visit method (although I don't really like that since it would match any `visit()` that isn't explicitly defined).  I'm not certain either of those are better than this macro.